ServerStreaming retries

igorbernstein2 opened the pull request:
Add retries and resume support. This is last part of #433

This PR consists of 3 parts split by commit:
- minor refactoring of test helpers to enable tests on the later parts
- Implement retries w. resume
- Implement support for timeouts

Conceptually, retries for streams can be split into 2 phases: before the first response is observed and after. Before the first response is observed, retries can be handled automatically similar to unary RPCs. After the first response is observed, manual work needs to done to recalculate the retry request. This work is delegated to a StreamTracker, which is notified of incoming responses and can be asked to build a resume request on failure. Gax should offer out of box support for the first kind of retries and allow client developers to manually add support for resumes via handwritten code. I envision the workflow for client developers as: add retry codes & settings to GAPIC yaml to enable simple retries, then implement a StreamTracker and set it on the ServerStreamingCallSettings.

This PR also implements timeouts for RPCs. The existing concepts of RPC & total timeouts are extended: RPC timeout now limits the time interval between a consumer signaling demand for the next response via StreamController#request() and receiving a response in ResponseObserver#onResponse. Total timeout limits the total duration of the stream from the initial call() until the last onComplete/onError of the last retry attempt. Furthermore, the concept of idle timeout is added to streams with manual flow control, this limits the time between last observed consumer activity (the time between finishing processing a response in onResponse to the next call to request). This is meant to address the possibility of the caller forgetting to cancel a partially read stream.

Since GRPC does not implement the concept of RPC timeouts for streams, this PR implements it as  a watchdog helper. The helper tracks activity per stream and schedules periodic garbage collection sweeps. So stream RPC timeouts are a lot looser than their unary counterparts. Furthermore, the helper needs to receive more information from its callers than is currently available: it needs to be able to accept both an rpc timeout and a total stream deadline. I couldn't figure out how to generalize this and extend the ApiContext to carry this information.  So the watchdog helper steps outside the callable chain pattern and is implemented as a factory for ResponseObserver decorators.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/Watchdog.java:
```diff
@@ -0,0 +1,316 @@
+      Preconditions.checkArgument(count > 0, "count must be > 0");
+      Preconditions.checkState(!autoAutoFlowControl, "Auto flow control is enabled");
+
+      // Only reset the request water mark if there is no outstanding requests.
```
s/if there is/if there are/

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/Watchdog.java:
```diff
@@ -0,0 +1,316 @@
+    @Override
+    public void onResponse(ResponseT response) {
+      synchronized (lock) {
+        state = State.DELIVERING;
```
Could a response be received in the middle of stale cancellation?

igorbernstein2 replied:
If I understand your concern correctly, no, the caller thread only sets the cancellation flag, the actual cancellation is plumbed through grpc by calling cancel() on the controller, so the actual cancellation will come from the same thread that would deliver the response and grpc has strict rules about callback order

igorbernstein2 replied:
I added #453 that makes sure that the rules are enforced for custom callables

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/retrying/StreamTracker.java:
```diff
@@ -0,0 +1,58 @@
+ * <p>Implementations don't have to be threadsafe because all of the calls will be serialized.
+ */
+@BetaApi("The surface for streaming is not stable yet and may change in the future.")
+public interface StreamTracker<RequestT, ResponseT> {
```
It doesn't feel like a factory, how about `StreamResumptionMediator`?

garrettjonesgoogle replied:
Hmm I don't feel like it is mediating either. How about `StreamResumptionStrategy` or `StreamResumptionAlgorithm`?

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/rpc/Callables.java:
```diff
@@ -75,9 +75,18 @@ private Callables() {}
+            clientContext.getClock(),
+            callSettings.getTimeoutCheckInterval(),
+            callSettings.getIdleTimeout());
+    watchdog.start();
```
Ideally not, but this was the least invasive approach. Another option would be work this into ClientContext, but I would prefer to hold off on doing this now

garrettjonesgoogle replied:
Ok, we can leave it as is and worry about optimizations later. Maybe add a comment about the implementation decision and future alternatives?

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/rpc/Watchdog.java:
```diff
@@ -0,0 +1,316 @@
+    DELIVERING
+  }
+
+  class Stream implements ResponseObserver<ResponseT> {
```
I see your point that WatchdogStream is a pretty large class. On the other hand, Watchdog is a tiny wrapper. Extracting WatchdogStream will create quite a bit of  noise:
-  WatchdogStream needs almost all of Watchdog's instance variables. Most of Watchdog's state exists only to be passed on to WatchdogStream. 
- There is bidirectional communication:
  - Watchdog needs to invoke stale checks
  - WatchdogStream needs to update a shared shared map owned by Watchdog

Furthermore, I'm not convinced that there is a lot of value to be gained from the split.
- I can't think of a situation where someone would want to use WatchdogStream outside of Watchdog. 
- Watchdog will just be an empty shell with no functionality of its own to test
- WatchdogStream will be just as testable.

This probably means that the 2 classes are too coupled, but decoupling them would mean introducing listeners or callback handles which would create a different kind of noise.

Let me know if you see a cleaner way to extract the inner class or if you think the benefit of extracting the inner class is worth the cost.

garrettjonesgoogle replied:
I don't have any suggestions so let's go ahead with it as an inner class.

igorbernstein2 commented:
PTAL, all feedback should be addressed

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/Callables.java:
```diff
@@ -75,6 +75,10 @@ private Callables() {}
 
+    // NOTE: This creates a Watchdog per streaming API method. Ideally, there should only be a
+    // single Watchdog for the entire process, however that change would be fairly invasive and
+    // the cost of multiple Watchdogs is fairly small, they all use the same executor. If this
```
s/small, they/small, since they/

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/StreamResumptionStrategy.java:
```diff
@@ -0,0 +1,58 @@
+import com.google.api.core.BetaApi;
+
+/**
+ * This is part of the server streaming retry api. It's implementers are responsible for tracking
```
s/It's/Its/

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/StreamResumptionStrategy.java:
```diff
@@ -0,0 +1,58 @@
+  /** Creates a new instance of this StreamResumptionStrategy without accumulated state */
+  StreamResumptionStrategy<RequestT, ResponseT> createNew();
+
+  /** Called by the {@link RetryingServerStream} to notify of a successfully received response. */
```
Better: `Called by the {@link RetryingServerStream} when a response has been successfully received.`

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/StreamResumptionStrategy.java:
```diff
@@ -0,0 +1,58 @@
+  /**
+   * Called when a stream needs to be restarted, the implementation should generate a request that
+   * will yield a stream whose first response would come right after the last response in
+   * onProgress.
```
s/yield a stream/yield a new stream/

s/after the last response in onProgress/after the last response received by onProgress/

This wording seems to indicate that SimpleStreamResumptionStrategy is not complying with the expectations described here.

garrettjonesgoogle commented:
@vam-google could you take a look too since you did the unary retry implementation?

igorbernstein2 commented:
All feedback addressed

igorbernstein2 commented:
rebased

vam-google commented on gax/src/main/java/com/google/api/gax/rpc/ServerStreamingCallSettings.java:
```diff
@@ -52,11 +113,120 @@ private ServerStreamingCallSettings(Builder<RequestT, ResponseT> builder) {}
+      this.retrySettings = RetrySettings.newBuilder().build();
+      this.resumptionStrategy = new SimpleStreamResumptionStrategy<>();
+
+      this.timeoutCheckInterval = Duration.ZERO;
```
It seems like `null` is allowed in setter methods (for this and other fields of this builder class, see `setTimeoutCheckInterval()` for example). In such case it seems leaving those field as `null` in case if default constructor called should be ok (and better). Another option would be to add validation on setters or (I strongly believe as the most preferable) add validation in (and ONLY in) constructed class  (`ServerStreamingCallSettings`). This would ensure that an invalid object cannot be constructed (regardless of how the constructor is called: via builder, directly, via reflection), and also will allow "invalid" intermediate states of a builder, which should be ok, because builders may collect data over time and do not have all needed information on the moment of creation of the builder itself (thus should allow "invalid" state until build() is called).

igorbernstein2 replied:
I added precondition checks on the setters. I think it's a better user experience to get the call stack of how the null entered the system

vam-google commented on gax/src/main/java/com/google/api/gax/rpc/Watchdog.java:
```diff
@@ -0,0 +1,314 @@
+      Duration checkInterval,
+      Duration idleTimeout) {
+
+    Preconditions.checkNotNull(executor, "executor can't be null");
```
Minor: throwing NullPointerException with stacktrace should be enough to figure out what was null. We usually do not have explicit error message for this, so `Preconditions.checkNotNull(executor)` should be just enough here (same for other non-null params).

igorbernstein2 replied:
I would prefer to keep it, it's not hurting anything and will save someone a trip to the source to figure out which param was null.

vam-google commented on gax/src/main/java/com/google/api/gax/rpc/Watchdog.java:
```diff
@@ -0,0 +1,314 @@
+    WatchdogStream(ResponseObserver<ResponseT> responseObserver, Duration waitTimeout) {
+      this.waitTimeout = waitTimeout;
+      this.outerResponseObserver = responseObserver;
+      this.lastActivityAt = clock.millisTime();
```
Minor: please move this initialization to L161 to make it consistent with the other `@GuardedBy` fields.

igorbernstein2 replied:
done

vam-google commented on gax/src/main/java/com/google/api/gax/rpc/Watchdog.java:
```diff
@@ -0,0 +1,314 @@
+        }
+
+        // Increment the request count without overflow
+        int maxIncrement = Integer.MAX_VALUE - pendingCount;
```
Same comment as for RetryingServerStream L165 (very suspicious math here).

vam-google commented on gax/src/main/java/com/google/api/gax/rpc/Watchdog.java:
```diff
@@ -0,0 +1,314 @@
+  }
+
+  /** The marker exception thrown when a timeout is exceeded. */
+  public static class IdleConnectionException extends ApiException {
```
Minor: please add `serialVersionUID` as this class is serializable. Also, should this class be public (maybe package-private?)

igorbernstein2 replied:
done

vam-google commented on gax/src/main/java/com/google/api/gax/rpc/Watchdog.java:
```diff
@@ -0,0 +1,314 @@
+    }
+  }
+
+  public static final StatusCode LOCAL_ABORTED_STATUS_CODE =
```
Minor: can this be package-private?

igorbernstein2 replied:
I would like to leave this public to allow users to catch this

igorbernstein2 commented on gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallableFactory.java:
```diff
@@ -235,7 +235,13 @@ private GrpcCallableFactory() {}
+        new GrpcExceptionServerStreamingCallable<>(
+            callable, streamingCallSettings.getRetryableCodes());
+
+    if (!streamingCallSettings.getRetryableCodes().isEmpty()
```
I don't mind making the change, but isn't this more consistent with `GrpcServerStreamingRequestParamCallable` above?

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/rpc/Callables.java:
```diff
@@ -65,6 +66,35 @@ private Callables() {}
+    // becomes an issue, the watchdog can be moved to ClientContext.
+    Watchdog<ResponseT> watchdog =
+        new Watchdog<>(
+            clientContext.getExecutor(),
```
The thing to keep in mind that this particular watchdog is meant to address 2 specific purposes:
1. to deal with the user forgetting to close ServerStream without fully consuming it
2. some bug in grpc that prevents the stream from being notified that it was closed.

So in that respect it's best effort and not designed to deal with timely clean up on a loaded system. In a perfect world, we would not need this, however support experience with the current bigtable client makes this a requirement.

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/rpc/Callables.java:
```diff
@@ -65,6 +66,35 @@ private Callables() {}
+            clientContext.getClock(),
+            callSettings.getTimeoutCheckInterval(),
+            callSettings.getIdleTimeout());
+    watchdog.start();
```
Eventually I would like to move the watchdog into its own ServerStreamingCallable. But that will require changes to ApiContext that I don't want to block on. So I would like to keep the watchdog as self contained as possible for the time being

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/rpc/RetryingServerStreamingCallable.java:
```diff
@@ -0,0 +1,83 @@
+/*
```
Done. But in the future, I think a better approach to address issues in already committed code would be to open an issue and assign it to me.

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/rpc/RetryingServerStreamingCallable.java:
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2018, Google LLC All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Done, with the same caveat as above

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/rpc/RetryingServerStreamingCallable.java:
```diff
@@ -0,0 +1,83 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
```
Please open a github issue where this can be discussed.

igorbernstein2 commented:
I've implemented most feedback, still outstanding on my part:
- factor out retry business logic into `RetryAlgorithm`, whose design is awaiting review
- factor a `RetryExecutor`-like from RetryStream

Things that I addressed, but don't think need code changes:
- watchdog threadpool 
- watchdog self start
- if Callables.retryabe should be responsible for excluding the callable

Discussion of pendingRequestCount and ServerStreamingCallable call method naming should be moved to a separate issue

igorbernstein2 commented:
Closing in favor of #463
